### PR TITLE
[fix] Implement pod draining

### PR DESCRIPTION
**Pull request by 0x416e746f6e**
This PR implements separate HTTP server listening on `DRAIN_ADDR`.  When it receives HTTP GET, it will flag server as un-healthy.  From that moment on `/health` endpoint of the main HTTP server will start reporting `HTTP 500`.

If we use this in conjunction with k8s pre-stop hook, we can make sure that on updates/restarts the pod lives long enough to serve whatever requests the load-balancer would still forward to it until it detects that the server is not healthy.

For example:

```yaml
          lifecycle:
            preStop:
              httpGet:
                port: 8081
          env:
            - name: DRAIN_ADDR
              value: :8081
```

### CODE REVIEW

**Comment by metachris** [comment]
Interesting! This PR does a lot and I don't have the resources to review the nuances of these changes right now. Will defer to @dvush and @Ruteri

File: server/server.go
Comment (0x416e746f6e) [inline server/server.go:78]: **explanation:**

move the routing setup into the `NewRpcEndPointServer` method.  this will make it possible to run more than one server (on different ports, for example).  also, it's nice to encapsulate the creation logic into one place.

**Comment by 0x416e746f6e** [comment]
oh, there's also https://github.com/flashbots/rpc-endpoint/pull/114 that somewhat overlaps.  sorry @dvush, I didn't see it before.  please let me know if I should refactor this PR to build on top of yours.

**Comment by dvush** [comment]
@0x416e746f6e , yes adapt please. It already handles graceful shutdown except health endpoint part

**Comment by 0x416e746f6e** [comment]
> yes adapt please. It already handles graceful shutdown except health endpoint part

@dvush:  updated!  please have a look.

File: cmd/server/main.go
Comment (0x416e746f6e) [inline cmd/server/main.go:97]: **explanation:**

instead of passing a train of arguments to the `NewRpcEndPointServer` (which would render it error-prone due to arguments being misplaced) let's use `struct` that will have all of them properly named.

File: server/server.go
Comment (dvush) [inline server/server.go:196]: Who will call this? Pre stop hook?

I thought that we only need to handle SIGTERM and then start returning errors on  /health

File: server/server.go
Comment (dvush) [inline server/server.go:200]: So the idea is that request to the drain endpoint should take 60 seconds?

File: server/server.go
Comment (dvush) [inline server/server.go:219]: don't we need some kind of mutex/atomic for this?

File: server/server.go
Comment (dvush) [inline server/server.go:106]: My understanding is that we don't need to have separate server for that and call "drain" method.

All we need is to: 
1. start returning errors from /health after this line is crossed
2. Wait n-seconds here instead of in the drain handler

If its the same then I would prefer to do it here with one server because its more simple and straightforward. 

But maybe its not the same?

Comment (0x416e746f6e) [reply server/server.go:106]: I also thought about that idea (sleep in the `SIGTERM` handler).

reason I did not go for it was that we only need to do that (sleeping) when we run it in k8s behind AWS LB target group (because it takes its time to become aware of the pod being drained).  while in the case of plain k8s service, the service becomes aware of the pod being drained basically immediately (I guess the control plane first updates the service's endpoints, and only after sends a signal to the pod).

if we introduce the sleep into SIGTERM handler then we'd impose that on all workflows regardless if that would be necessary or not.

File: server/server.go
Comment (0x416e746f6e) [reply server/server.go:200]: yes.

reason is that control plane considers the pre-hook to be "done" right at the moment when the request returns a status, which is when it will send the SIGTERM to the pod.

File: server/server.go
Comment (0x416e746f6e) [reply server/server.go:219]: in the purest sense, maybe "yes".  but practically, not really.  we don't mind if there's race-condition here (if not this health-check sees the new status, then the next one will).  and it's ok if there is concurrent access (we don't play with pointers here, and we read the value of isHealthy only once in this method).

File: server/server.go
Comment (dvush) [reply server/server.go:106]: I mean something like this https://github.com/flashbots/rpc-endpoint/pull/116

File: server/server.go
Comment (0x416e746f6e) [reply server/server.go:219]: but maybe worthwhile to put this into a comment inside the source-code.  who knows what other functionality will be added here in the future, so the future maintainer should better be aware.

File: server/server.go
Comment (dvush) [reply server/server.go:106]: You probably right, I agree. Did not see your comment

**Review by dvush — APPROVED** [review]

**Pull request merged**
